### DISCUSSION TITLE
Possibility to use third-party compiled library binaries of ffmpeg (gyan.dev)

### DISCUSSION THREAD

**Discussion by kivisade** [asticode/go-astiav#78] [Q&A] [answered]
Hi! @asticode! Thank you for providing this excellent library!

I would like to ask for advice regarding possible use of third-party built binaries of `ffmpeg` (in particular, @GyanD's build of the [release 7.0](https://github.com/GyanD/codexffmpeg/releases/tag/7.0)) with your library, instead of building `ffmpeg` locally.

## Background of the question

I am building on Windows. I am not very familiar with building Golang applications requiring C library dependencies, but I was able to set up [MinGW-w64](https://www.mingw-w64.org/) and compile ffmpeg using your instructions provided in [`README.md`](https://github.com/asticode/go-astiav?tab=readme-ov-file#install-ffmpeg-from-source) and [`Makefile`](https://github.com/asticode/go-astiav/blob/master/Makefile#L9) (although had to face a resolve an issue described [here](https://trac.ffmpeg.org/ticket/6620)). After that, I was able to build my Golang app using your library and confirm it working for my use case. However, the build process is tedious and time-consuming, and I am not sure that I am knowledgeable enough to make decisions regarding necessary options for `./configure`. If possible, I would like to re-use third party builds such as https://www.gyan.dev/ffmpeg/builds/ to avoid the complexity of setting up the build environment and running the build process myself.

After some experimentation, I was also able to build an application using your library and the binary files from @GyanD ([`ffmpeg-7.0-full_build-shared.7z`](https://github.com/GyanD/codexffmpeg/releases/download/7.0/ffmpeg-7.0-full_build-shared.7z)). 

However, the main issue with this approach is that @GyanD's binary distribution **does not include `pkgconfig` files** which are necessary to build the Golang app with your library.

I was able to circumvent this problem by taking the `lib/pkgconfig` files which were generated during the local build, and placing them in the same path `lib/pkgconfig` under the directory with the extracted [`ffmpeg-7.0-full_build-shared.7z`](https://github.com/GyanD/codexffmpeg/releases/download/7.0/ffmpeg-7.0-full_build-shared.7z).

I understand that the problem with this approach is that it is not future-compatible, i.e. when you decide to update the version of ffmpeg supported by your library, new `pkgconfig` files would need to be generated to be able to use third-party complied binaries.

## Question

I'm sorry for a vague question, but I would like to ask if there could be any better solution to allow using your library with vendored binaries? 

I would greatly appreciate your advice 🙏

**Comment by kivisade** [comment]
For reference, I would like to share the Makefiles that I came up with, which support both building `ffmpeg` from source locally as described in your [`README.md`](https://github.com/asticode/go-astiav?tab=readme-ov-file#install-ffmpeg-from-source) and [`Makefile`](https://github.com/asticode/go-astiav/blob/master/Makefile#L9), and using downloaded files from @GyanD's [release 7.0](https://github.com/GyanD/codexffmpeg/releases/tag/7.0).

(I am quoting only relevant parts, omitting some of the variable declarations which are not relevant to subject)

# Usage

```bash
# Local build
make build-ffmpeg

# Download and prepare built binary files from gyan.dev
make get-gyan.dev-ffmpeg

# Build app with locally build ffmpeg libs
make FFMPEG_VER=n7.0 rebuild

# Build app with ffmpeg libs from gyan.dev
make rebuild
```

# Code

## `Makefile` at root of project

```Makefile
FFMPEG_VER  := gyan.dev-7.0
FFMPEG_ROOT := ffmpeg/$(FFMPEG_VER)

# Environment variables related to building with ffmpeg dependency,
# see https://github.com/asticode/go-astiav?tab=readme-ov-file#install-ffmpeg-from-source
BUILD_ENV_VARS := CGO_ENABLED=1 \
                  CGO_LDFLAGS="-L$$(realpath $(FFMPEG_ROOT)/lib)" \
                  CGO_CFLAGS="-I$$(realpath $(FFMPEG_ROOT)/include)" \
                  PKG_CONFIG_PATH="$$(realpath $(FFMPEG_ROOT)/lib/pkgconfig)"

$(BIN_FILE):
	go mod tidy
	$(BUILD_ENV_VARS) go build -o $(BIN_FILE) $(LDFLAGS) ./cmd/$(COMMAND_NAME)

.PHONY: clean-bin
clean-bin:
	rm -f $(BIN_FILE)

.PHONY: rebuild
rebuild: clean-bin $(BIN_FILE)

.PHONY: build-ffmpeg get-gyan.dev-ffmpeg
build-ffmpeg get-gyan.dev-ffmpeg:
	$(MAKE) -C ffmpeg $@
```

## `ffmpeg/Makefile`

```Makefile
#==================================================================================
# Option 1: Building from source @ https://github.com/FFmpeg/FFmpeg
#----------------------------------------------------------------------------------

FFMPEG_BUILD_VER         := n7.0
FFMPEG_REPO              := https://github.com/FFmpeg/FFmpeg
FFMPEG_SRC_PATH          := $(FFMPEG_BUILD_VER)/src
FFMPEG_PATCH             := $(shell realpath "patch/ffbuild-library.mak.patch")
FFMPEG_CONFIGURE_OPTIONS := --enable-static --disable-shared

# ffmpeg build instructions based on https://github.com/asticode/go-astiav/blob/master/Makefile
.PHONY: build-ffmpeg
build-ffmpeg:
	rm -rf $(FFMPEG_SRC_PATH)
	mkdir -p $(FFMPEG_SRC_PATH)
	git clone --branch $(FFMPEG_BUILD_VER) --single-branch --depth 1 $(FFMPEG_REPO) $(FFMPEG_SRC_PATH)
ifneq "" "$(FFMPEG_PATCH)"
	cd $(FFMPEG_SRC_PATH) && git apply $(FFMPEG_PATCH)
endif
	cd $(FFMPEG_SRC_PATH) && ./configure --prefix=.. $(FFMPEG_CONFIGURE_OPTIONS)
	cd $(FFMPEG_SRC_PATH) && make
	cd $(FFMPEG_SRC_PATH) && make install
	rm -rf $(FFMPEG_SRC_PATH)

#==================================================================================
# Option 2: Downloading from https://github.com/GyanD/codexffmpeg/releases/tag/7.0
#----------------------------------------------------------------------------------

FFMPEG_GYAND_VER   := 7.0
FFMPEG_GYAND_ROOT  := gyan.dev-$(FFMPEG_GYAND_VER)
GYAN_DEV_DIST_FILE := ffmpeg-$(FFMPEG_GYAND_VER)-full_build-shared.7z
GYAN_DEV_DIST_URL  := https://github.com/GyanD/codexffmpeg/releases/download/$(FFMPEG_GYAND_VER)/$(GYAN_DEV_DIST_FILE)

.PHONY: get-gyan.dev-ffmpeg
get-gyan.dev-ffmpeg:
	rm -rf $(FFMPEG_GYAND_ROOT)
	mkdir -p $(FFMPEG_GYAND_ROOT)
	wget -q -P $(FFMPEG_GYAND_ROOT) $(GYAN_DEV_DIST_URL)
	cd $(FFMPEG_GYAND_ROOT) && bsdtar --strip-components=1 -xf $(GYAN_DEV_DIST_FILE)
	rm $(FFMPEG_GYAND_ROOT)/$(GYAN_DEV_DIST_FILE)
	cp -r pkgconfig $(FFMPEG_GYAND_ROOT)/lib/
```

## `ffmpeg/patch/ffbuild-library.mak.patch`

```
diff --git a/ffbuild/library.mak b/ffbuild/library.mak
index 793e9d41fa..1104819e4a 100644
--- a/ffbuild/library.mak
+++ b/ffbuild/library.mak
@@ -35,7 +35,9 @@ OBJS += $(SHLIBOBJS)
 endif
 $(SUBDIR)$(LIBNAME): $(OBJS) $(STLIBOBJS)
 	$(RM) $@
-	$(AR) $(ARFLAGS) $(AR_O) $^
+	$(file >objs.txt,$^)
+	cat objs.txt
+	$(AR) $(ARFLAGS) $(AR_O) @objs.txt
 	$(RANLIB) $@
 
 install-headers: install-lib$(NAME)-headers install-lib$(NAME)-pkgconfig
```

## Directory `ffmpeg/pkgconfig`

Contains `*.pc` files generated by local build under `ffmpeg/n7.0/lib/pkgconfig` when running `make build-ffmpeg`.

**Comment by asticode** [comment, marked as answer]
Thing is, from `go-astiav`'s standpoint, using `pkgconfig` is the best way since it removes all the complexity of gathering all proper `CFLAGS` and `LDFLAGS` for `ffmpeg` depending on the OS.

Therefore unfortunately I don't see another option for you than providing the `pkgconfig` files 🤔 Apart from creating an issue in gyan.dev's repo to get `pkgconfig` files in the releases.

Also, if you build manually ffmpeg to get the `pkgconfig` files, why use the gyan.dev release? 🤔

**Comment by GyanD** [comment]
I'll consider including pc files with the next release. However, the prefix will need to be adjusted by the user.

@asticode  The rationale for a vendor build may be the third-party libraries linked. My full build referred to here has 80 external libs.
